sklearn/utils/_array_api.py

```diff
@@ -595,3 +597,52 @@ def _estimator_with_converted_arrays(estimator, converter):
 def _atol_for_type(dtype):
     """Return the absolute tolerance for a given dtype."""
     return numpy.finfo(dtype).eps * 100
+
+
+def _average(X, axis=None, weights=None):
+    """Compute the weighted average along the specified axis.
+
+    This function is a port of numpy.average that supports the Array API.
+
+    Please see the original documentation for more details:
+    https://numpy.org/doc/stable/reference/generated/numpy.average.html
+
+    Parameters
+    ----------
+    X : array object
+        Array containing data to be averaged.
+    axis : None or int or tuple of ints, optional
+        Axis or axes along which to average `X`.
+    weights : array_like, optional
+        An array of weights associated with the values in `X`. Each value in
+        `X` contributes to the average according to its associated weight.
+
+    Returns
+    -------
+    retval : array object
+        Return the average along the specified axis.
+    """
+    xp, _ = get_namespace(X)
+
+    if _is_numpy_namespace(xp):
+        return xp.asarray(numpy.average(X, axis=axis, weights=weights))
+
+    if weights is None:
+        return xp.mean(X, axis=axis)
+
+    weights = xp.asarray(weights, device=device(X))
+    if X.shape != weights.shape:
+        if axis is None:
+            raise TypeError(
+                "Axis must be specified when shapes of a and weights differ."
```

Comment:
```suggestion
                "Axis must be specified when shapes of X and weights differ. "
                f"Got {X.shape = } and {weights.shape = }."
```

Reply:
Please add test for this error message and the following. Please also include the observed shape values in the error message when they are not valid.